### DISCUSSION TITLE
fetch MessageType given its interface?

### DISCUSSION THREAD

**Discussion by shawnjones253** [timostamm/protobuf-ts#651] [Q&A]
I'm creating a schema registry, with the initial goal of simply validating that objects match their respective proto definition before being serialized and written anywhere.

Given that a developer would start with:
```
const data: ExampleMessage = {
  myString: 'abc',
  myInt: 1,
}
```

I'd like to implement a utility function that calls the right `toJson()` and uses some project-wide `JsonWriteOptions`:
```
export function messageToJson<T extends object>(
  message: T,
  messageType: IMessageType<T>,
  options?: Partial<Omit<JsonWriteOptions, 'emitDefaultValues'>>
) {
  return messageType.toJson(message, { ...options, emitDefaultValues: true })
}
```

Is there any way to get rid of the `messageType: IMessageType<T>` parameter here and infer it from the message?

**Comment by timostamm** [comment]
This is generally not possible in TypeScript, since all type information is erased at runtime.

**Comment by jcready** [comment]
You _can_ get this to work, but only for objects (and nested objects) that were created via the protobuf-ts classes (e.g. `SomeMessage.create()`, `SomeMessage.fromBinary()`, etc.). Basically all message objects output by the protobuf-ts classes are "branded" with a non-enumerable property whose value is the `MessageType` class which created it. So you can either keep your existing function and have the caller find the `MessageType` before calling `messageToJson`:

```ts
import { containsMessageType, MESSAGE_TYPE } from '@protobuf-ts/runtime';

//... somehow acquire `data` object

if (containsMessageType(data)) {
  const messageType = data[MESSAGE_TYPE];
  console.log(messageToJson(data, messageType));
}
```

Or put the logic into the `messageToJson` function with the caveat that the first argument might not have a `MessageType`. 

```ts
import { containsMessageType, MESSAGE_TYPE } from '@protobuf-ts/runtime';

export function messageToJson<T extends object>(
  message: T,
  options?: Partial<Omit<JsonWriteOptions, 'emitDefaultValues'>>
) {
  if (containsMessageType(message)) {
    const messageType = message[MESSAGE_TYPE];
    return messageType.toJson(message, { ...options, emitDefaultValues: true });
  }
  throw new Error('No MessageType on provided input');
}
```
